Move from `setup.py` to `pyproject.toml`; Update build scripts

aaraney opened the pull request:
Move from `setup.py` to `pyproject.toml`.
Sub-package authorship is also updated.
Build scripts are also updated.

Replaces #683.
Closes #683.

Related to:
- #680
- #248
- #247 

Closes #248
Closes #247

robertbartel commented on python/lib/communication/pyproject.toml:
```diff
@@ -0,0 +1,37 @@
+    "ngen-config@git+https://github.com/noaa-owp/ngen-cal@master#egg=ngen-config&subdirectory=python/ngen_conf",
+]
+readme = "README.md"
+description = "Communications library package for components of the National Water Model as a Service architecture"
```
Let's take the opportunity to change this also.

```suggestion
description = "Communications library package for components of the DMOD architecture"
```

robertbartel commented on python/lib/core/pyproject.toml:
```diff
@@ -0,0 +1,26 @@
+]
+dependencies = ["pydantic>=1.10.8,~=1.10", "typing_extensions"]
+readme = "README.md"
+description = ""
```
Might as well add something here too.

```suggestion
description = "Library package with core DMOD concrete and abstract types depended upon by other DMOD libraries."
```

robertbartel commented on python/lib/evaluations/pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+    { name = "Robert Bartel" },
+]
+dependencies = [
+    "dmod.core",
```
This may not be the place to address it, but it is probably a bug that no version constraint for `dmod.core` is specified here (likely also true for the next line and `dmod.metrics`).

robertbartel commented on python/lib/modeldata/pyproject.toml:
```diff
@@ -0,0 +1,42 @@
+    "pyogrio",
+]
+readme = "README.md"
+description = ""
```
Might as well put something here too.

```suggestion
description = "Library package for classes related to forcing data, metadata, and other modeling-related data types and operations needed for tasks that can be executed within DMOD."
```

robertbartel commented on python/lib/scheduler/pyproject.toml:
```diff
@@ -0,0 +1,39 @@
+    "pydantic>=1.10.8,~=1.10",
+]
+readme = "README.md"
+description = ""
```
And this.

```suggestion
description = "Library packages defining types for jobs, resources, execution operations, and related support components."
```

robertbartel commented on python/services/dataservice/pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+    "ngen-cal@git+https://github.com/noaa-owp/ngen-cal@master#egg=ngen-config&subdirectory=python/ngen_cal",
+]
+readme = "README.md"
+description = ""
```
```suggestion
description = "Service package for data management service within DMOD architecture."
```

robertbartel commented on python/services/requestservice/pyproject.toml:
```diff
@@ -0,0 +1,33 @@
+    "dmod.externalrequests>=0.6.0",
+]
+readme = "README.md"
+description = ""
```
```suggestion
description = "Service package for service component receiving and handling external request within the DMOD architecture."
```

robertbartel commented on python/services/schedulerservice/pyproject.toml:
```diff
@@ -0,0 +1,30 @@
+    "dmod.scheduler>=0.13.0",
+]
+readme = "README.md"
+description = ""
```
```suggestion
description = "Service package for service responsible for managing job scheduling, execution, and resource management in the DMOD architecture."
```

robertbartel reviewed (changes requested):
@aaraney, this look great!.  I did make a few inline suggestions to add `description` values in some of the .toml files where it was empty.  I also changed one `description` to remove a "MaaS" reference and replace with "DMOD," which is the only thing I think really needs to be changed.

aaraney commented:
@robertbartel, thanks for adding package descriptions! Great teamwork!

Just added all of your suggestions, @robertbartel, and squashed them so the history is a little cleaner and rebased off of master. Can you give it a quick re-review and get this on the road?
